### ISSUE TITLE
Training details

### ISSUE THREAD

**Issue by buaaeducn** [yyfz/Pi3#31]
Dear devs,
Thanks for opensourceing such great work! I tried to use a new dataset on the mdoel and used the loss from the paper, but the loss does not converge. How do you train it? Could you tell me some details? (like epoches, time...)

**Comment by yyfz** [comment]
Hi, [buaaeducn](https://github.com/buaaeducn)!

This is probably becuase of the implementation of the training loss. Could you share me the loss you use?

**Comment by buaaeducn** [comment]
> Hi, [buaaeducn](https://github.com/buaaeducn)!  嗨，[buaaeducn](https://github.com/buaaeducn)！
> 
> This is probably becuase of the implementation of the training loss. Could you share me the loss you use?这可能是因为实施了训练损失。你能分享一下你使用的损失吗？

Becauese my final goal is to get the relatively accurate camera_pose, so I just use the camera_pose_loss from the paper.
```
def camera_pose_loss(pred_poses, gt_poses, rot_weight=1.0, trans_weight=1.0, eps=1e-8, delta=1.0):
    """改进的相机姿态损失函数：相对姿态监督 + 旋转测地距离 + 平移Huber损失"""
    # 提取旋转矩阵 (B, T, N, 3, 3) 和平移向量 (B, T, N, 3)
    R_pred = pred_poses[..., :3, :3]
    R_gt = gt_poses[..., :3, :3]
    t_pred = pred_poses[..., :3, 3]
    t_gt = gt_poses[..., :3, 3]
    
    # 1. 相对姿态监督（消除全局参考模糊）
    R_pred_rel = torch.matmul(R_pred[..., 1:, :, :], R_pred[..., :-1, :, :].transpose(-2, -1))
    R_gt_rel = torch.matmul(R_gt[..., 1:, :, :], R_gt[..., :-1, :, :].transpose(-2, -1))
    
    t_pred_rel = t_pred[..., 1:, :] - torch.matmul(R_pred_rel, t_pred[..., :-1, :, None]).squeeze(-1)
    t_gt_rel = t_gt[..., 1:, :] - torch.matmul(R_gt_rel, t_gt[..., :-1, :, None]).squeeze(-1)
    
    # 2. 旋转测地距离损失
    trace = R_pred_rel.matmul(R_gt_rel.transpose(-2, -1)).diagonal(dim1=-2, dim2=-1).sum(-1)
    angle = torch.acos(torch.clamp((trace - 1) / 2, -1 + eps, 1 - eps))
    rot_loss = angle.mean()
    
    # 3. 平移Huber损失
    trans_diff = t_pred_rel - t_gt_rel
    trans_norm = torch.norm(trans_diff, p=2, dim=-1)
    
    trans_loss_huber = torch.where(
        trans_norm <= delta,
        0.5 * trans_norm ** 2,
        delta * (trans_norm - 0.5 * delta)
    )
    trans_loss = trans_loss_huber.mean()
    
    total_loss = rot_weight * rot_loss + trans_weight * trans_loss
    return total_loss, rot_loss, trans_loss
def train_one_epoch(model, train_loader, optimizer, device, epoch, writer, 
                    rot_weight=1.0, trans_weight=0.1, rank=0):
    """训练一个epoch"""
    model.train()
    total_pose_loss = 0.0
    total_rot_loss = 0.0
    total_trans_loss = 0.0
    batch_count = 0
    
    # 只有主进程显示进度条
    iterator = tqdm(train_loader, desc=f"[Train] Epoch {epoch}") if rank == 0 else train_loader
    
    for step, batch in enumerate(iterator):
        imgs = batch['image'].to(device)
        gt_extrinsics = batch['extrinsics'].to(device)  # [B, T, N, 4, 4]
        B, T, N, C, H, W = imgs.shape
        
        optimizer.zero_grad()
        
        # 收集T帧的预测结果
        batch_camera_poses = []
        for i in range(T):
            img = imgs[:, i]  # [B, N, C, H, W]
            results = model(img)
            batch_camera_poses.append(results['camera_poses'])  # [B, N, 4, 4]
        pred_poses = torch.stack(batch_camera_poses, dim=1)  # [B, T, N, 4, 4]
        
        # 计算损失
        pose_loss, rot_loss, trans_loss = camera_pose_loss(
            pred_poses=pred_poses,
            gt_poses=gt_extrinsics,
            rot_weight=rot_weight,
            trans_weight=trans_weight
        )
        
        # 反向传播和优化
        pose_loss.backward()
        optimizer.step()
        
        # 累计损失
        total_pose_loss += pose_loss.item()
        total_rot_loss += rot_loss.item()
        total_trans_loss += trans_loss.item()
        batch_count += 1
        
        # 只有主进程记录TensorBoard
        if rank == 0 and step % 10 == 0:
            global_step = epoch * len(train_loader) + step
            writer.add_scalar("Train/Loss_pose", pose_loss.item(), global_step)
            writer.add_scalar("Train/Loss_rot", rot_loss.item(), global_step)
            writer.add_scalar("Train/Loss_trans", trans_loss.item(), global_step)
    
    # 计算平均损失
    avg_pose = total_pose_loss / batch_count
    avg_rot = total_rot_loss / batch_count
    avg_trans = total_trans_loss / batch_count
    
    # 只有主进程记录每个epoch的平均损失
    if rank == 0:
        writer.add_scalar("Train/Avg_Loss_pose", avg_pose, epoch)
        writer.add_scalar("Train/Avg_Loss_rot", avg_rot, epoch)
        writer.add_scalar("Train/Avg_Loss_trans", avg_trans, epoch)
    
    return avg_pose, avg_rot, avg_trans
```

**Comment by buaaeducn** [comment]
<img width="1513" height="866" alt="Image" src="https://github.com/user-attachments/assets/a65a891a-835a-4c7d-bf41-3d64f5a67273" />

**Comment by yyfz** [comment]
Please make sure the the `pred_pose` and `gt_pose` is c2w if you are using above implementation. If they are w2c, the relative pose you calculated is wrong.

**Comment by yyfz** [comment]
> > Please make sure the the `pred_pose` and `gt_pose` is c2w if you are using above implementation. If they are w2c, the relative pose you calculated is wrong.如果您使用上述实现，请确保 `pred_pose` 和 `gt_pose` 为 c2w。如果它们是 w2c，则您计算的相对姿势是错误的。
> 
> I check again, and I am sure that pred_pose and gt_pose is c2w. I am confused with this problem...

Oh, the relative pose should compute every pairs. You just compute the relative poses of 'fram i to frame i+1'.

**Comment by yyfz** [comment]
> I would be very grateful if you could share the specific relative camera loss implementation used for training. I've been attempting to replicate this, but my model's predicted camera poses are resulting in NaN values shortly after training begins. Any guidance you could offer would be greatly appreciated.

Try these:
1. Clip the input to torch.acos to be within [−1+ϵ,1−ϵ].
2. Do not train from scratch, as it underperforms the VGGT baseline when only predicting camera poses and local pointmaps. Instead, load the pre-trained VGGT alternating attention module weights for better results.

**Comment by wenzhoulyu** [comment]
> > I would be very grateful if you could share the specific relative camera loss implementation used for training. I've been attempting to replicate this, but my model's predicted camera poses are resulting in NaN values shortly after training begins. Any guidance you could offer would be greatly appreciated.
> 
> Try these:
> 
> 1. Clip the input to torch.acos to be within [−1+ϵ,1−ϵ].
> 2. Do not train from scratch, as it underperforms the VGGT baseline when only predicting camera poses and local pointmaps. Instead, load the pre-trained VGGT alternating attention module weights for better results.

Thank you for your advice. I suspect the instability stems from the unstable gradient of `torch.acos`. Using `torch.atan2` might be a more stable approach.
